Async Git Repository class

thedaniel opened the pull request:
This is an experimental spike of an alternate GitRepository class that uses [nodegit](http://www.nodegit.org/) rather than [git-utils](https://github.com/atom/git-utils). We would love to speed up the responsiveness of various aspects of Atom (tree-view, etc) by doing more Git operations asynchronously, and it would also be nice to get out of the business of maintaining our own libgit2 wrapper.

Here are a few notes about the reasoning behind some decisions here:
- It follows the existing GitRepository API closely, changing everything sync to Promise-based async.
  - The assumption is that having a close mapping between the sync and aysnc APIs will make it less painful to refactor client packages.
  - I'm not 100% sure we should keep this design, so it's a tradeoff between design quality and incurred labor. Ideally we'll keep this private and experimental all the way until Atom 2.0 so that we can change it as we need to.
- The class is written in ES6, because Coffeescript, while lovely, is the past, and the community seems to have consensus on transpilers like babel.
  - This includes an .eslintrc, because I've found it extremely useful.
  - The spec is still in coffee because I thought it might be easier to just modify it as I go to make sure I had the same coverage on both classes, though at the moment I wish I'd run it through a coffee -> js converter first.
- I haven't done a lot of ES6 or constructed a Promise-based API for anything serious before, so I'm sure to misuse some APIs.
- We'll need to keep the sync GitRepository class around for backwards compat, so I haven't decided quite how to integrate this with the current provider system.
  - Perhaps we add a private method on the sync repo like `GitRepository.async` that returns the Promise for an async repo, so that internal callers can use async without changing anything in the `GitRepositoryProvider`?
- Documentation is sparse because I wasn't sure this was a good path to follow until today. I'll flesh it out more as I go.

cc @atom/core @orderedlist 
- [ ] Determine a good test case (even if manual / subjective) for git-related UI responsiveness
  - many simultaneous changes, check tree-view update?
  - simulate degraded disk performance / network drive?
- [ ] cf #9240

nathansobo commented:
What's the plan for transitioning to this API? I was thinking our existing synchronous `GitRepository` could always have an `async` property that points to this new version. We can then encourage users of today's APIs to use the async version like this:

```js
let repository = someMethodReturningARepository().async
```

When we hit Atom 2.0 and transition to the repository *always* being async, we can retain the `async` property and just point it to self, such that `repository.async === repository` returns `true`.

However, I'm not sure how this all meshes with your idea of keeping this API private for a while. In my mind, we should start transitioning the community to any async APIs we feel confident about. We might not want to make every method on the class public, but I'd like to deprecate existing synchronous functionality as soon as we can.

thedaniel commented:
>  I was thinking our existing synchronous GitRepository could have an async property that points to this new version. 

Yep, I agree with this, it's suggested in the bullet points above.

> However, I'm not sure how this all meshes with your idea of keeping this API private for a while. In my mind, we should start transitioning the community to any async APIs we feel confident about. 

This is not necessarily a conflict with how I feel, I'm just not expecting we'll feel confident about the APIs for a while.

nathansobo commented on spec/git-repository-async-spec.coffee:
```diff
@@ -0,0 +1,406 @@
+
+      onSuccess = jasmine.createSpy('onSuccess')
+
+      waitsForPromise ->
```
If you also write the specs in ES6, `async/await` syntax could really clean up these async tests...

```js
it('does something async', function () {
  waitsForPromise(async function() {
    expect(await repo.getPath()).toBe('/some/path')
  })
})
```

thedaniel replied:
I think that's ES7+ only

thomasjo replied:
Yes, which we have access to :smile:

nathansobo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,191 @@
+  }
+
+  constructor (path) {
+    this.repo = null
```
What's this used for?

nathansobo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,191 @@
+
+  isPathIgnored (_path) {
+    return this.repoPromise.then((repo) => {
+      return Promise.resolve(Git.Ignore.pathIsIgnored(repo, _path))
```
Is `pathIsIgnored` synchronous?

nathansobo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,191 @@
+
+  isPathModified (_path) {
+    return this._filterStatusesByPath(_path).then(function (statuses) {
+      var ret = statuses.filter((status) => {
```
V8 supports the new ES6 `Array.prototype.every` method, which would be clearer and save you allocation of a new array.

nathansobo commented:
:zap: Thanks for moving us in the right direction on this.

nathansobo commented:
> it's suggested in the bullet points above.

Sorry, I didn't comprehend it initially for some reason. Of course `async` will have to be a promise, and when we transition all our repository-returning APIs to be asynchronous, we'll have to have an `async` property on the returned *promise* that references the original promise.

smashwilson commented:
Oh hey, this will be excellent for packages that integrate with git. :+1: :grin:

alexandernst commented:
:heart: ES6 instead of :coffee:script!

thedaniel commented:
Thought: it seems like a good place to start using this in prod & packages is to proxy all the `did-change-status` events from the sync implementation to the async implementation. At this moment I'm just writing tests that assert the events fire correctly in async-land at all and calling through from sync to async in a few methods (e.g. `getPathStatus(path)`) just so the side-effects happen, but ultimately we could remove anything solely concerned with generating events out of the sync implementation entirely and just re-emit the ones generated by the async implementation.

philipgiuliani commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,272 @@
+    return new GitRepositoryAsync(path, options)
+  }
+
+  constructor (path, options) {
```
What do you think about removing the spaces in front of the arguments? https://github.com/airbnb/javascript#18.3

thedaniel replied:
Good catch - this currently conforms to [JS Standard Style](https://github.com/feross/standard) (a very presumptuous name, isn't it?) rather than AirBnB style and I'm happy with Standard.

philipgiuliani replied:
Thanks for the resource. Didn't knew about it! Looks of course also good :+1: Just important that there is one style over the whole repository :+1: :clap:

steelbrain commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,255 @@
+
+    var {project, refreshOnWindowFocus} = options
+    this.project = project
+    if (refreshOnWindowFocus === undefined) {
```
If you want a boolean value from `refreshOnWindowFocus`, then use `refreshOnWindowFocus = Boolean(refreshOnWindowFocus)`. It'll turn `false`, `null` and others to `false` and `true` remains the same.

thedaniel replied:
I actually want to check undefined here, because it should default to `true` - that is to say, `true` is `true` and `undefined` is `true`, but only explicitly set`false` is `false`, which isn't quite how the Boolean constructor works.

steelbrain commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,255 @@
+  }
+
+  getPath () {
+    return this.repoPromise.then((repo) => {
```
in ES6, if you are only getting one parameter, you can write `(repo) => { }` as just `repo => { }`

steelbrain commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,255 @@
+      var ret = statuses.filter((status) => {
+        return status.isModified()
+      }).length > 0
+      return Promise.resolve(ret)
```
You don't necessarily have to return a promise from `Promise.prototype.then`

steelbrain commented on .eslintrc:
```diff
@@ -0,0 +1,6 @@
+{
+  "env": {
```
May I suggest the eslint [`no-var`](http://eslint.org/docs/rules/no-var.html) clause?

thedaniel replied:
That's not a bad idea. There's a bunch of other JS in the repo that won't comply to this, but we don't need to have CI fail on eslint failures right away..

thedaniel commented:
Thanks for the feedback @steelbrain!

steelbrain commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+
+  getPath () {
+    return this.repoPromise.then((repo) => {
+      return Promise.resolve(repo.path().replace(/\/$/, ''))
```
You might want to replace most of these `Promise.resolve`s here with just values.

peduxe commented:
You guys really hate placing semicolons don't you? :D

SeanJM commented:
By not adding semi colons, are you adding overhead to the JIT?

> On Oct 23, 2015, at 12:04 PM, Alex Cruz <notifications@github.com> wrote:
> 
> You guys really hate placing semicolons don't you? :D
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/atom/atom/pull/9213#issuecomment-150619013>.
>

peduxe commented:
@SeanJM I don't know why github doesn't use them in their code, is this because they have most of their code in Ruby?

SeanJM commented:
https://www.quora.com/Are-there-performance-differences-between-using-semi-colons-and-not-using-semi-colons-in-JavaScript <https://www.quora.com/Are-there-performance-differences-between-using-semi-colons-and-not-using-semi-colons-in-JavaScript>
According to this the answer is no.

> On Oct 23, 2015, at 1:11 PM, Alex Cruz <notifications@github.com> wrote:
> 
> @SeanJM <https://github.com/SeanJM> I don't know why github doesn't use them in their code, is this because they have most of their code in Ruby?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/atom/atom/pull/9213#issuecomment-150636085>.
>

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+    this._gitUtilsRepo = GitUtils.open(path) // TODO remove after porting ::relativize
+    this.repoPromise = Git.Repository.open(path)
+
+    var {project, refreshOnWindowFocus} = options
```
This should be `let` (or `const`), doesn't look like we want or need hoisting here.

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+  isPathModified (_path) {
+    return this._filterStatusesByPath(_path).then(function (statuses) {
+      return statuses.filter((status) => {
+        return status.isModified()
```
This can be changed into
```js
return statuses.filter(status => status.isModified())
```

You should probably also be able to rewrite the whole block as
```js
return this._filterStatusesByPath(_path).then(statuses => statuses.filter(status => status.isModified()))
```

thedaniel replied:
I had been trying to only use `=>` where I needed lexical this (though it snuck in here), is it typical to just use it everywhere?

thomasjo replied:
They are typically used everywhere. I use them just as I use skinny arrows in CS.

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+  isPathNew (_path) {
+    return this._filterStatusesByPath(_path).then(function (statuses) {
+      var ret = statuses.filter((status) => {
+        return status.isNew()
```
Same deal here with the `var`, and collapsing the single return function.

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+
+  checkoutHead (_path) {
+    return this.repoPromise.then((repo) => {
+      var checkoutOptions = new Git.CheckoutOptions()
```
`let`

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+      checkoutOptions.checkoutStrategy = Git.Checkout.STRATEGY.FORCE | Git.Checkout.STRATEGY.DISABLE_PATHSPEC_MATCH
+      return Git.Checkout.head(repo, checkoutOptions)
+    }).then(() => {
+      return this.getPathStatus(_path)
```
Can be collapsed.

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+
+  checkoutHeadForEditor (editor) {
+    return new Promise(function (resolve, reject) {
+      var filePath = editor.getPath()
```
`let`

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+        reject()
+      }
+    }).then((filePath) => {
+      return this.checkoutHead(filePath)
```
Can be collapsed.

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+  // Returns a Promise that resolves to the status bit of a given path if it has
+  // one, otherwise 'current'.
+  getPathStatus (_path) {
+    var relativePath = this._gitUtilsRepo.relativize(_path)
```
`let`

thomasjo commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,271 @@
+      this.pathStatusCache[relativePath] = status
+      return Promise.resolve(status)
+    })
+  }
```
`let`s, and collapses :smiley:

thedaniel commented:
Hm, I only have 2 (seemingly unrelated) failures on my machine, but there are a bunch of :red_circle: on travis. In any case this is starting to come together nicely.

nathansobo commented:
> but there are a bunch of :red_circle: on travis

You may want to run tests against the current released version rather than master locally just in case that's a factor.

thedaniel commented:
Whoops! @nathansobo i meant to post that comment on the tree-view PR. Tabs are hard.

thedaniel committed:
Start async spec over in JS

thedaniel committed:
Start adding the troublesome buffer-events first

thedaniel committed:
add status-changed on reload spec

thedaniel committed:
Flesh out ::destroy()

thedaniel committed:
Add standardjs config to package.json

thedaniel committed:
Use async’s destroy callback in repo provider.

thedaniel committed:
Destroy async repo from sync repo

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Get these tests going.

joshaber committed:
Async/await a lot of things.

joshaber committed:
Async harder.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Cite it.

thedaniel commented:
Over in #9469 as of this moment I have just the buffer callback stuff implemented (i.e. the specs that are weirdly flaky on this branch) and consistently green. I'm not sure what the essential difference is though. One difference is that I implemented `::relativize` in that PR rather than just proxying to the git-utils version, but I'm not sure how that's relevant. I also implemented the specs slightly differently, with some gentle use of async/await. In any case, tomorrow I'll port some of the differences over to this branch and see if it makes things green here, else I'll port things from here to there one section of specs at a time and see if any one area starts to make async-land flaky.

joshaber commented:
So I *think* this branch is in a pretty good spot now.

1. Ported more tests to JS.
2. Moved from Jasmine waiting to async/await as much as I know how.
3. Fixed a race between file writing/git reading.

The tests _seem_ to consistently pass locally now. I'd be really curious to know how they behave for you now too :pray: 

My plan for Monday was to:

- [ ] Continue porting tests to JS.
- [ ] Make `terribleWait` and `terribleSleep` less terrible.
- [ ] See about getting CI building again so we can have some consistency to testing.

If you wanted, we could divide and conquer that.

thedaniel commented:
> Fixed a race between file writing/git reading.

😻

> If you wanted, we could divide and conquer that.

Sure thing, let me attack this CI thing first.

thedaniel committed:
Merge branch 'master' into dh-async-repo

thedaniel commented:
The last build failed on a bunch of license checks, at least a couple related to eslint & standard added as `devDependencies`. One could argue that those libs are only used to develop, not in the distribution, so it's fine that their licenses aren't permissive. though I'm not 100% sure that holds water. For now I'll just take them out of the package.json (but leave in the config for standard) and see if that clears up this check license task.

thedaniel committed:
remove devDependencies due to license issues

thedaniel committed:
add license overrides for shelljs and log-driver

thedaniel committed:
Add empty ignore array to standard config

- This version of grunt-standard requires SOMETHING to be set here, or it dies
  hard because it tries to push undefined into the array of ignored paths

thedaniel committed:
Clean up lint errors from `./script/grunt standard:src`

thedaniel committed:
coffeelint fix

thedaniel committed:
unforce spec

thedaniel committed:
add final buffer event spec

joshaber committed:
It's better if we actually run the tests.

joshaber committed:
Merge remote-tracking branch 'origin/dh-async-repo' into dh-async-repo

thedaniel committed:
Copy relativize and getPathStatus from #9469

joshaber committed:
Use asyncIt here too.

thedaniel committed:
Use our own ::relativize

This allows us to remove the git-utils import.

thedaniel committed:
Merge branch 'dh-async-repo' of github.com:atom/atom into dh-async-repo

thedaniel committed:
await getCachedPathStatus in specs

joshaber committed:
:fire: terribleWait.

thedaniel committed:
Revert "Use our own ::relativize"

This reverts commit 0322d8d3dada0334edba74768ce4b54725ca7757.

thedaniel committed:
Revert "await getCachedPathStatus in specs"

This reverts commit d8985c8175f25e6f40026c07980846c68d4da935.

joshaber committed:
Focus async its.

joshaber committed:
Wait for the first refresh to complete.

joshaber committed:
Cleanup

joshaber committed:
Unfocus

joshaber committed:
Re-enable that test.

joshaber committed:
Re-add this test.

joshaber committed:
Use node-pre-gyp to grab binaries of nodegit.

joshaber commented:
So the tests are in a pretty good spot now, I think. There's still a potential race condition but I haven't seen in practice. I'm thinking over how to best fix it.

But the more pressing problem is that CI is timing out.

`master` currently builds in ~1200s, where this is timing out at 1800s. So somehow this added ~10 minutes to the build.

I'm not entirely sure *where* the time's being spent. Anecdotally we seem to spend a lot of time in the `Installing apm...` phase. We're using `node-pre-gyp` to download the nodegit binary (or at least, that happens locally; unclear how to verify that from CI output).

@atom/feedback Anyone have ideas? If not I'll probably try to break some of the steps down and time them.

orderedlist commented:
nodegit takes forever to build in apm because there aren't any binaries. This is probably something we should fix in nodegit.

thedaniel commented:
Also, atom.io can have high latency sometimes, and we hit it many times to install core packages. I do plan to do a performance pass there at some point in the future, but in any case it would be so cool if we could just bundle / cache those dependencies rather than going over the network.

joshaber commented:
>nodegit takes forever to build in apm because there aren't any binaries. This is probably something we should fix in nodegit.

nodegit uses node-pre-gyp to provide binaries, but it looks like it special-cases Electron apps to always build from source: https://github.com/nodegit/nodegit/blob/b259b0587c6e44e3939e49055b90845fbfce9221/lifecycleScripts/install.js#L27 Presumably because Electron wasn't supported until fairly recently? https://github.com/mapbox/node-pre-gyp/pull/175

>but in any case it would be so cool if we could just bundle / cache those dependencies rather than going over the network

I'd love to do some sort of binary caching but I'd really love for it to not block this.

thedaniel commented:
@joshaber I brought up the `Installing apm...` thing in chat the other day because it takes forever on my laptop as well though it "should be fast", worth looking into. Though it would be a shame to yak shave too much. The nodegit node-gyp path seems like the lowest-hanging fruit.

joshaber committed:
Expose the promise for refreshStatus.

joshaber committed:
Be even less racy.

joshaber committed:
Merge branch 'master' into dh-async-repo

thedaniel committed:
Add simple ::relativize()

- remove spec
- Remove GitUtils from async repo
- Doesn't cover all the cases as the one in git-utils but I think they might have
  been specific to that library's implementation

thedaniel committed:
Merge branch 'dh-async-repo' of github.com:atom/atom into dh-async-repo

orderedlist commented:
Installing nodegit actually has to compile libgit2 as well, which takes some time. Realistically, an `apm` binary for nodegit would make this wayyyyy faster.

joshaber commented:
Yeah I'm looking into adding Electron binaries to nodegit.

thedaniel committed:
Use async repo, add missing return

thedaniel committed:
i am bad at using GH desktop

thedaniel committed:
Fix typo

joshaber committed:
We don't need node-pre-gyp as a dependency.

thedaniel committed:
Add cache config from @joefitzgerald's branch.

If this doesn't speed us up enough we'll cache node_modules as well :godmode:

thedaniel committed:
Merge branch 'dh-async-repo' of github.com:atom/atom into dh-async-repo

johnhaley81 commented:
@orderedlist is right with building libgit2 but it's even worse than that. NodeGit has to build libgit2/libssh2/openssl which is as horrible as you think it is. Luckily we don't have a C++11 requirement anymore but it's still super slow.

Anyways, so nodegit/nodegit#783 is in and ready to go so really the next step is to have the CI build/publish binaries for linux/darwin/win32 during a release. That way everything is just automated and ready to go. The only thing that would break is if a new version of electron is released there's no way for nodegit to pick up on that and we won't be able to grab a binary for the latest.

The only way around that (that I can think of right now) is just having a box that hooks into releases of electron and automates building/publishing for nodegit.

This might not be too much of a problem when Electron gets to v1 but still, it's going to be a pain to always go back and build/publish this guy when Electron gets a version bump.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Don't need these, we'll use the JS one.

joshaber committed:
Implement the Windows focus refreshing.

joshaber committed:
ES6 harder.

joshaber commented:
:metal: Tests are passing, just need to get build times down so it doesn't time out :metal:

maxbrunsfeld commented on spec/git-repository-async-spec.js:
```diff
@@ -0,0 +1,457 @@
+  xit(name, () => {
+    waitsForPromise(fn)
+  })
+}
```
@nathansobo ended up adding functions like these in [async-spec-helpers.coffee](https://github.com/atom/atom/blob/master/spec/async-spec-helpers.coffee). They detect if a promise is returned from the `it` function, so you can just shadow `it` and friends.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Const and comparison fixes.

h/t @YuriSolovyov

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
git-repository will refresh us on window focus.

joshaber committed:
Update the current branch as well.

joshaber committed:
Use modern imports.

joshaber committed:
Gimme the linters.

joshaber committed:
These are fine.

joshaber committed:
Remove some lint.

joshaber committed:
ES6 some more.

joshaber committed:
License overrides.

joshaber committed:
These can be const too.

joshaber committed:
More const'ing.

joshaber committed:
Update the overrides.

joshaber committed:
Make sure we're testing the right thing.

joshaber committed:
Let's use a separate method for side-effecting.

joshaber committed:
Call destroy on the synchronous repo.

joshaber committed:
Ensure the repo is destroyed properly after testing.

joshaber committed:
100% less racy.

joshaber committed:
Don't log anymore.

joshaber committed:
Use refreshStatusForBuffer.

joshaber committed:
:fire: all the disabled tests.

joshaber committed:
These may be called more than once and that's ok.

joshaber committed:
Do the work after waiting.

joshaber committed:
Of course it's JS.

joshaber committed:
Call the better-named method.

joshaber committed:
Take refreshStatusForPath into account for _refreshingCount.

joshaber committed:
Use the ES6 export syntax.

joshaber committed:
Add back window refreshing.

joshaber committed:
Organize similarly to git-repository.coffee so we can more easily tell what we're missing.

joshaber committed:
Added getShortHead

joshaber committed:
Added getAheadBehindCount.

joshaber committed:
Added getDiffStats.

joshaber committed:
More helpful type annotations.

joshaber committed:
Move the refreshingCount changes closer to where the work is done.

joshaber committed:
Don't make the async repo duplicate the work if it's wrapped by a sync repo.

joshaber committed:
Added .hasBranch

joshaber committed:
Added .getUpstreamBranch

joshaber committed:
Added .getReferences

joshaber committed:
Added .getReferenceTarget

joshaber committed:
Added .getConfigValue

joshaber committed:
A lil more efficient .getReferenceTarget.

joshaber committed:
Added .getOriginalURL.

joshaber committed:
Add some more stubs/TODOs.

joshaber committed:
Added .checkoutReference

joshaber committed:
Added .getLineDiffs

joshaber committed:
Set IGNORE_WHITESPACE_EOL on Windows.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Use some binaries, hopefully?

joshaber committed:
Added license override for inherit.

joshaber committed:
Use nodegit-atom for now.

joshaber committed:
Don't set these here.

joshaber committed:
Do the needful to get Electron binaries.

joshaber committed:
Use the same version as The Real Nodegit.

joshaber committed:
++nodegit

joshaber committed:
++nodegit

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Provide these from .npmrc.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber commented on .npmrc:
```diff
@@ -1 +1,4 @@
 cache = ~/.atom/.npm
+runtime = electron
+target = 0.34.5
```
I don't like that we have to put this here and in package.json. Anyone have ideas about how we could not do that?

joshaber replied:
Maybe set the environment variable from within bootstrap? Or something?

50Wliu commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,719 @@
+          // Ignore eol of line differences on windows so that files checked in
+          // as LF don't report every line modified when the text contains CRLF
+          // endings.
+          options.flags = Git.Diff.OPTION.IGNORE_WHITESPACE_EOL
```
:heart_eyes: Will this close #9424?

joshaber replied:
Huh... Unfortunately I doubt it. Our [current implementation](https://github.com/atom/atom/blob/77a540962312ecf2a9cedd483cb4329564590458/src/git-repository.coffee#L387) is supposed to use this flag too.

joshaber committed:
Get the Electron version from the package.

joshaber committed:
Back to putting them in the npmrc.

joshaber committed:
It's fine if there isn't an upstream.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Just use the repo declared above.

joshaber committed:
Fix usage when we're given no options.

joshaber committed:
Tests for .checkoutHeadForEditor.

joshaber committed:
Throw if we try to use the repository after it's been destroyed.

joshaber committed:
Test .destroy().

joshaber committed:
Match the ordering of the GitRepository spec.

joshaber committed:
Add a whole ton of missing documentation.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Delint

joshaber commented:
Alright, I believe this is ready for review :pray: @atom/feedback 

There are two caveats:

1. The Windows builds are still timing out because we're having issues getting Windows nodegit binaries. Still a work in progress. I won't merge this until that's taken care of.
2. Submodule support is incomplete. I'm [working](https://github.com/nodegit/nodegit/pull/821) [on](https://github.com/nodegit/nodegit/pull/825) it, but I think this can land without it, so long as everyone's aware.

My current plan is:

1. Add submodule support once a new nodegit version is out.
2. Convert `tree-view`, `status-bar`, and `git-diff` to async.

kevinsawicki commented on spec/git-repository-async-spec.js:
```diff
@@ -0,0 +1,598 @@
+
+    it('resolves true for an ignored path', async () => {
+      const ignored = await repo.isPathIgnored('a.txt')
+      expect(ignored).toBeTruthy()
```
Might be worth making these a little "stronger" with `toBe(true)` since this API should always return a `boolean` right? Looks like the `isStatusModified` specs do that currently.

joshaber replied:
That's a really great point :+1:

kevinsawicki commented on spec/git-repository-async-spec.js:
```diff
@@ -0,0 +1,598 @@
+      let threw = false
+      try {
+        await repo.getShortHead()
+      } catch (e) {
```
Would it be better to catch this exception and assert the message and/or type? To ensure it throws the expected error and not just any error.

kevinsawicki commented on spec/git-repository-async-spec.js:
```diff
@@ -0,0 +1,598 @@
+  })
+
+  describe('buffer events', () => {
+    let repository
```
Super minor, but it looks like most of the other times in other specs this variable is called `repo` and not `repository`, not sure if it would be worth using a consistent name throughout, but just wanted to mention it.

kevinsawicki commented on spec/git-repository-async-spec.js:
```diff
@@ -0,0 +1,598 @@
+      editor.insertNewline()
+
+      let called
+      repository.onDidChangeStatus(c => called = c)
```
I think it would be better to use a Jasmine spy here and wait on the `callCount` and then assert on  `toHaveBeenCalledWith` helper.

kevinsawicki commented on spec/git-repository-async-spec.js:
```diff
@@ -0,0 +1,598 @@
+      repo = GitRepositoryAsync.open(workingDirectory)
+    })
+
+    it('can get the line diff', async () => {
```
Maybe something like "returns the old and new lines of the diff" would be a clearer name for this spec.

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,870 @@
+        // TODO: This'll probably be wrong if the submodule doesn't exist in the
+        // index yet? Is that a thing?
+        const entry = index.getByPath(_path)
+        const submoduleMode = 57344 // TODO compose this from libgit2 constants
```
This seems like a constant that might make sense to put at the top with the status flags.

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,870 @@
+    }
+  }
+
+  // Public: Destroy this {GitRepositoryAsync} object.
```
Minor question, does our documentation parser work for JavaScript docs already?

joshaber replied:
I don't think so. I think that's what https://github.com/atom/donna/issues/9 is about.

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,870 @@
+  //
+  // * `directoryPath` The {String} path to check.
+  //
+  // Returns a promise resolving to a {Number} representing the status. This value can be passed to
```
`promise` -> `{Promise}`

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,870 @@
+  // Note that if the status of the path has changed, this will emit a
+  // 'did-change-status' event.
+  //
+  // path    :: String
```
This appears to be in a different documentation format than the other methods in this file.

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,870 @@
+    )
+
+    this.subscriptions.add(bufferSubscriptions)
+    return
```
I don't think this explicit `return` is needed since it is the last line of this function.

kevinsawicki commented on src/git-repository.coffee:
```diff
@@ -75,6 +76,15 @@ class GitRepository
       throw new Error("No Git repository found searching path: #{path}")
 
+    asyncOptions = {}
+    for key, val of options
```
Super minor, but this could be `asyncOptions = _.clone(options)` instead.

kevinsawicki commented:
> Alright, I believe this is ready for review :pray: /cc @atom/feedback

Gave it a review pass, left some comments. I know a lot of this code was ported over from the original sync version so I apologize if any of the comments pertain to things that were legacy and are just showing up in the diff because of the movement.

maxbrunsfeld commented:
:heart: This looks great.

joshaber committed:
We want the booleans.

joshaber committed:
Ensure we actually do give them a boolean.

joshaber committed:
Be more specific in our error names.

joshaber committed:
Use the usual name.

joshaber committed:
Tinker tailor soldier spy.

joshaber committed:
Better test name.

joshaber committed:
Put this constant with the others.

joshaber committed:
Fix the documentation style.

joshaber committed:
:doughnut: need this.

joshaber committed:
Just Use Clone.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber commented:
@kevinsawicki Thanks! :sparkling_heart: I think I :hammer:'d them all.

Any other feedback is welcome. Otherwise this is just waiting on Windows nodegit binaries.

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,871 @@
+    // XXX _filterSBD already gets repoPromise
+    return this.repoPromise
+      .then(repo => {
+        relativePath = this.relativize(directoryPath, repo.workdir())
```
This variable appears to be only used in this promise but is declared outside this scope on line 402, could the `let` move here instead?

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,871 @@
+
+  // Public: Returns a {Promise} that resolves to true if at the root, false if
+  // in a subfolder of the repository.
+  isProjectAtRoot () {
```
Looks like this method returns `null` instead of a `Promise` if `this.project` is falsey. Should it return a rejected Promise in that case instead?

kevinsawicki replied:
Or maybe in that case it should return a promise that just resolves to false.

joshaber replied:
Yup, good catch :+1:

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,871 @@
+  // Returns a {Boolean} that's true if the `statusBit` indicates the path is
+  // ignored.
+  isStatusIgnored (statusBit) {
+    return (statusBit & (1 << 14)) > 0
```
Super minor, worth pulling `1 << 14` to a `ignoredStatusFlags` constant for consistency?

kevinsawicki commented:
> Any other feedback is welcome.

Took another pass, left a few minor comments. This is looking really good though :+1:

joshaber committed:
This can be local.

joshaber committed:
Define a constant for this.

joshaber committed:
If we don't have a project then we're not at the root.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber commented:
:zap: :eyes:

kevinsawicki commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,873 @@
+    if (this.projectAtRoot) {
+      return this.projectAtRoot
+    } else {
+      this.projectAtRoot = Promise.resolve(() => {
```
Looks like `this.projectAtRoot` won't be returned when this case is hit since there isn't a return below here.

joshaber replied:
:astonished:

joshaber committed:
And that's why we write tests.

joshaber committed:
Linter

YuriSolovyov commented on spec/git-repository-async-spec.js:
```diff
@@ -0,0 +1,620 @@
+}
+
+function copyRepository () {
+  let workingDirPath = temp.mkdirSync('atom-working-dir')
```
I guess can be `const` ?

YuriSolovyov commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,873 @@
+    }
+
+    if (process.platform === 'win32') {
+      _path = _path.replace(/\\/g, '/')
```
Can't we make use of `node`s core libs like `path` here?

joshaber replied:
That's a good question. As the comment indicates, this was cargo-culted from `git-utils` so I don't have any context on it. @thedaniel any idea?

YuriSolovyov commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,873 @@
+  // is modified.
+  isPathModified (_path) {
+    return this._filterStatusesByPath(_path).then(statuses => {
+      return statuses.filter(status => status.isModified()).length > 0
```
Can we use [Array.prototype.some](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/some)
```javascript
return statuses.some(status => status.isModified())
```
here?

The thing is that it can return earlier without iterating whole collection and also not producing new array.

joshaber replied:
Fantastic :metal:

YuriSolovyov commented on src/git-repository-async.js:
```diff
@@ -0,0 +1,873 @@
+            filteredBits.forEach(bit => directoryStatus |= bit)
+          }
+
+          return directoryStatus
```
how about:
```javascript
return bits.filter(b => b > 0).reduce((status, bit) => { return status |= bit }, 0);
```
?

joshaber committed:
Const.

joshaber committed:
Use `some` instead.

joshaber committed:
Use the functions.

joshaber commented:
Thanks @YuriSolovyov!

joshaber committed:
If there isn't an index entry then it's not a submodule.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Be consistent in our path relativization.

joshaber commented:
Look at that lovely green checkmark :heart_eyes:

joshaber committed:
Fix hunk callback position.

joshaber committed:
Accumulate the raw data we want instead of keeping hunks around.

joshaber committed:
#nocontext

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Fix the test.

joshaber committed:
Try running core specs on appveyor.

joshaber committed:
Enable tests on windows.

joshaber committed:
Revert "Enable tests on windows."

This reverts commit d98bbb1119fef552b182476e0e31019d047e031b.

joshaber committed:
Revert "Try running core specs on appveyor."

This reverts commit 1c01d05440e42c84c300a80066c126ae03f0effa.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Always set the architecture to ia32 on Windows.

joshaber commented:
I've verified Mac and Windows binaries are working. Linux will have to wait until after the new year.

thedaniel commented:
Coincidentally, 2016 is the year of Linux on the desktop, so that works out great.

joshaber commented:
:penguin: :fire:

joshaber commented:
Even though it's not yet 2016, I verified that the Linux build is working :metal: :penguin:

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Provide messages for all these steps.

joshaber committed:
Get the Electron version from our package.json.

joshaber committed:
Syntax

joshaber committed:
How about this.

joshaber committed:
Just pass through the environment.

joshaber committed:
These commands will log, so no need to do it ourselves.

joshaber committed:
Not an object anymore.

joshaber committed:
Try passing them as args.

joshaber committed:
Back to using npmrc.

joshaber committed:
Once more, with feeling.

joshaber committed:
Log the environment.

joshaber committed:
Try args again.

joshaber committed:
Try just setting the target?

joshaber committed:
Comment why.

joshaber committed:
Remove items that don't have any changes.

joshaber committed:
Added getCachedPathStatuses()

joshaber committed:
Fix path standardization on case-sensitive file systems.

joshaber committed:
Strip any leading /private/.

I don’t love that we have to do this manually, but I also can’t find
any node function that’ll do it for us :(

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Use .openExt instead of .open.

Use openExt to open the repository so we get the same repo-searching
behavior we had with git-utils.

joshaber committed:
Do proper path-filtering for status.

joshaber committed:
Only refresh the status for the open project.

joshaber committed:
Fix the openExt call.

joshaber committed:
_destroyed => _isDestroyed

joshaber committed:
We might not have a project.

joshaber committed:
Use a default parameter.

joshaber committed:
Relativize these paths too.

joshaber committed:
So much depends upon a return statement.

joshaber committed:
Use .getRepo everywhere.

joshaber committed:
Catch errors that are thrown while we're refreshing.

joshaber committed:
We don't actually use `repo`.

joshaber committed:
Documentation for .DestroyedErrorName

joshaber committed:
Added .openNodeGitRepository

joshaber commented:
Hoping to merge this tomorrow :pray:

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Commas are important.

joshaber commented:
There's reportedly a problem with the Linux nodegit binaries.

From @johnhaley81:

>It seems to only affect LTS linux...
>It's very strange. Like it's executing a CPU instruction that's not available to older versions of ubuntu.

So I'm going to hold off on merging this until that's resolved.

johnhaley81 commented:
I'll keep you posted as I find out more.

joshaber committed:
Merge branch 'master' into dh-async-repo

alexandernst commented:
CPU instructions are not bound to software, but to hardware. If Linux binaries are not working on LTS Ubuntu (as there is no such thing as LTS Linux) it means you're using unavailable libraries (most probably). `ldd` should be enough to check that.

joshaber commented:
FWIW, it worked fine for me in testing on Ubuntu 64-bit 14.0.4.3

johnhaley81 commented:
The LTS Linux was a typo. I meant Ubuntu. The issue was when requiring the NodeGit library it would throw a `SIGILL`. It throws this when it's initializing OpenSSL for the library but not at a specific line of code. It dies when it returns from the first method in the call stack after `init` is called.

This happens with the prebuilt binary off of Travis. I'm attempting to configure a VM to allow @joshaber to have a consistent repo since it seems to be a heisenbug as well.

alexandernst commented:
`SIGILL` errors are expected in OpenSSL. Check https://www.openssl.org/docs/faq.html#PROG17

johnhaley81 commented:
That says under debugger situations which we aren't in. We're in a full production build without an attached debugger. Is that expected under that scenario as well?

johnhaley81 commented:
Ahh I re-read it and it's saying that during it's query for a CPU's capabilities it could get a SIGILL. Running the code though it was happening during a request to `dev/random` and it did successfully get a result from that. It dies when it pops the function from the call stack.

joshaber committed:
s//openNodeGitRepository/openRepository.

joshaber committed:
path is never used.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
:arrow_up: nodegit@0.6.3

johnhaley81 commented:
@alexandernst here's the `ldd -v` output for each binary. The shared libs looks the same to me. Also, are those hex values the address in the binary where the lib is loaded? 

https://gist.github.com/johnhaley81/617f36595b35513cc2a4

joshaber committed:
s/upstreamByPath/upstream

We’ll track submodule status separately.

joshaber committed:
We'll store them by path instead of by name.

alexandernst commented:
@johnhaley81 Can we get the backtrace too?

johnhaley81 commented:
@alexandernst https://gist.github.com/johnhaley81/abbc6dd801eca780ae7f

alexandernst commented:
@johnhaley81 It looks either like nodegit was compiled for the wrong arch (can you show me how is it compiled?) or an OpenSSL bug (less likely, but possible).

johnhaley81 commented:
@alexandernst this is the prebuilt binary compilation log: https://travis-ci.org/nodegit/electron-npg-automator/jobs/97494428

johnhaley81 commented:
Also so you know, I have a "magic" Ubuntu box that's configured the same way that works. Clearly not everything is the same but it's a clean install, no updates, only those packages are install, same env-variables and same compiler and compiler version. And that works. I can give you any outputs from that machine that you like.

alexandernst commented:
@johnhaley81 Just to make sure I understand the situation. The bug you're getting exist only in a VM that you use to test? Or can you repro the bug easily in any LTS Ubuntu?

johnhaley81 commented:
@alexandernst I've had reports of users experiencing this across multiple versions of Ubuntu. I have only been able to repro this consistently on this on 2 VMs and only when running Ubuntu. Mint and CentOS work.

johnhaley81 commented:
Also my local install of Ubuntu works fine with either build.

johnhaley81 commented:
@alexandernst @joshaber mentioned that the [disassmble](https://gist.github.com/johnhaley81/cae460ffee3c5876a236) of the running app might help

joshaber commented:
```
(gdb) x/i $pc
=> 0x7fc83ef3851c <lh_new+332>:    vxorps %xmm0,%xmm0,%xmm0
```

This seems to be the instruction it's SIGILL'ing on.

alexandernst commented:
@johnhaley81 @joshaber http://www.felixcloutier.com/x86/XORPS.html

Do the users's CPUs that are reporting the problem support AVX?
Also, I'm yet to see how nodegit is exactly compiled (the link @johnhaley81 posted doesn't have that info). Are you compiling with `-march=native` or something?

johnhaley81 commented:
@alexandernst that travis box loads the repo from github and then we have the GYP file loaded https://github.com/nodegit/nodegit/blob/master/vendor/libgit2.gyp which sets the flags. That is how the makefile is generated. We then use `node-gyp` to compile the module.

johnhaley81 commented:
@alexandernst as far as AVX support, I'll get back to you on that. Looking into it as much as I can now.

alexandernst commented:
Maybe something weird is going on in the Makefile. Can we get a copy of the generated Makefile?

About the AVX, check if available with `cat /proc/cpuinfo| grep avx`

alexandernst commented:
Also, the version of clang, if possible.

johnhaley81 commented:
The definitely are lacking avx. So here's a potential fix: https://github.com/nodegit/electron-npg-automator/commit/d4a8725b7d42011cd39edf938d1927f97500eb04

The `use_obsolete_assembly` should force OpenSSL to NOT use that but I think that setting it to true might have not set the flag correctly? Trying this out now.

alexandernst commented:
There you go, it's a compiler flag. That's why it was working on your machine (which I guess it does support AVX) and not in the machines of some of your users. Find how to force the compilation process to avoid using unsupported CPU ops.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Refresh the status for submodules too.

joshaber committed:
Stop logging.

joshaber committed:
Added a fixture that has submodules.

joshaber committed:
Test it.

joshaber committed:
Use Grown Up repositories for submodules.

joshaber committed:
Make .getRepo() submodule-aware.

joshaber committed:
Update . getCachedUpstreamAheadBehindCount()

Take submodules into account.

joshaber committed:
More tests for the submodule cases.

joshaber committed:
Test .isSubmodule.

joshaber committed:
Just Use Or

joshaber committed:
:arrow_up: grunt-standard@2.0.0

So that it knows about async functions.

joshaber committed:
Linter was mad about these now.

joshaber committed:
:arrow_up: babel-eslint@5.0.0-beta6

joshaber committed:
Merge pull request #10327 from atom/submodules

Async submodule support

joshaber committed:
Merge branch 'master' into dh-async-repo

johnhaley81 commented:
Compiler flag didn't fix. Turns out it's possibly a bug in OpenSSL. Potential fix: https://github.com/nodegit/nodegit/pull/855

alexandernst commented:
Indeed, turns out it's an OpenSSL bug :)
https://github.com/openssl/openssl/commit/817ddb9fb07e78b451e838a77b2b272b0dd23e5f

johnhaley81 commented:
Confirmed, that fixes it.

thedaniel committed:
Keep track of openedPath when relativizing

.. otherwise if the repo root is a symlink, paths won’t relativize to
it and git statuses won’t be updated/shown.

thedaniel committed:
Always relativize to working directory

thedaniel committed:
Make linter happy

joshaber committed:
:arrow_up: node-git@0.7.0

johnhaley81 commented:
Bumping to NodeGit 0.7.0 should resolve that linux issue and make this PR gtg :)

joshaber committed:
Merge branch 'master' into dh-async-repo

thedaniel commented:
Hm, I guess I didn't cross-reference this here: https://github.com/atom/atom/pull/10352

^ This is a fix for ::relativize on this branch that still needs a little polish

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Serialize refreshing.

joshaber committed:
:fire: refreshingCount.

joshaber committed:
Use the promise from refreshStatus to know when we're done.

joshaber committed:
Log refresh failures.

thedaniel committed:
We don't need to add another param here, actually

joshaber commented:
I'm seeing intermittent failures about file watching, e.g.:

```
>> GitRepositoryAsync
>>   .checkoutHeadForEditor(editor)
>>     it does not display a dialog when confirmation is disabled
>>       Error: Leaking subscriptions for paths: /private/var/folders/gw/_2jq29095y7b__wtby9dg_5h0000gn/T/atom-working-dir116011-49688-dfmojy/a.txt
>>         at ensureNoPathSubscriptions (/Users/travis/build/atom/atom/spec/spec-helper.coffee:122:15)
>>         at [object Object].<anonymous> (/Users/travis/build/atom/atom/spec/spec-helper.coffee:115:3)
>>         at process._tickCallback (node.js:368:9)
>>   .destroy()
>>     it throws an exception when any method is called after it is called
>>       Error: Leaking subscriptions for paths: /private/var/folders/gw/_2jq29095y7b__wtby9dg_5h0000gn/T/atom-working-dir116011-49688-1pr3ez3/a.txt
>>         at ensureNoPathSubscriptions (/Users/travis/build/atom/atom/spec/spec-helper.coffee:122:15)
>>         at [object Object].<anonymous> (/Users/travis/build/atom/atom/spec/spec-helper.coffee:115:3)
>> 
>> 
```

`GitRepositoryAsync` itself doesn't do any file watching directly. Is this a bug somewhere else? Some cleanup we're not doing? @atom/core

joshaber committed:
Make sure the editor disappears quickly.

joshaber committed:
Capture less in .checkoutHeadForEditor.

joshaber committed:
Don't use require.resolve.

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber commented:
I moved some stuff around which maybe fixed that? I dunno.

I think we're just waiting on #10352, and then :tada: :balloon: :birthday:

thedaniel committed:
Add specs to cover ::relativizeToWorkingDirectory

Yay, this uncovered a problem with the implementation.

thedaniel committed:
be more concise, again

thedaniel committed:
Merge remote-tracking branch 'remotes/origin/dh-async-repo' into dh-async-repo-relativize-symlinks

thedaniel committed:
Annotate waitsFor in a couple flaky specs

thedaniel committed:
Add a few failure messages to waitsFors in flaky specs

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber committed:
Merge branch 'master' into dh-async-repo

thedaniel committed:
Merge remote-tracking branch 'remotes/origin/dh-async-repo' into dh-async-repo-relativize-symlinks

thedaniel committed:
Add option for custom label to waitsForPromise

thedaniel committed:
Don't let label be null

joshaber committed:
Try using .getStatusExt instead of .getStatus.

joshaber committed:
:fire: some lint.

joshaber committed:
Merge pull request #10352 from atom/dh-async-repo-relativize-symlinks

Support symlinked repo roots when relativizing

joshaber committed:
Merge branch 'master' into dh-async-repo

joshaber commented:
![](http://www.reactiongifs.com/r/wvy1.gif)

kevinsawicki commented:
Awesome work here :+1: 

![](https://49.media.tumblr.com/de2f235161b951996d7e27b89df45e99/tumblr_n9o0bqbST91rpc5kho1_500.gif)

johnhaley81 commented:
w00t!

![image](https://cloud.githubusercontent.com/assets/867780/12365767/222f4a46-bb94-11e5-9528-907b9a8b4f14.png)

martindale commented:
Does this address #2456?

50Wliu commented:
@martindale it would be great if you could download the 1.7.0 beta and find out :).

martindale commented:
Subjectively, the problem appears to be worse with 1.7.0 beta.  Atom now completely hangs when launching it inside of a remote mount, and eventually requests direction on whether to close or keep waiting.

yannickcr commented:
Same problem here, Atom 1.6.0 or 1.7.0 beta now hangs up when trying to open a big git repository inside of a remote mount making Atom unusable in this configuration :cry:

bynite commented:
Can't use Atom anymore. Probably because of this feature. I mounted a vagrant via SSHFS and Atom freezes when opening a file. Could you release a fix for this, please.

mehcode commented:
I'm wondering why we switched to using nodegit. I realize that maintaining a wrapper around a git executable is lame but nodegit wraps libgit2 which is glacial in comparison to git. This is completely anecdotal but I've always found it far slower.

Perhaps using something like https://www.npmjs.com/package/simple-git instead?

thedaniel commented:
@mehcode Before nodegit we were already using libgit2, with a sync api wrapper we were maintaining ourselves (https://github.com/atom/git-utils) - so we can't blame any performance regressions on libgit2 here.

mehcode commented:
Ah. I was under the impression we were just doing shell execution.
